internal/state/change_processor.go

```diff
@@ -60,23 +77,93 @@ type ChangeProcessorConfig struct {
 
 // ChangeProcessorImpl is an implementation of ChangeProcessor.
 type ChangeProcessorImpl struct {
-	store *store
-	cfg   ChangeProcessorConfig
+	// clusterState holds the current state of the cluster
+	clusterState graph.ClusterState
+	// updater acts upon the cluster state.
+	updater Updater
+	// getAndResetClusterStateChanged tells if the cluster state has changed.
+	getAndResetClusterStateChanged func() bool
 
-	// changed is true if any changes that were captured require an update to nginx.
-	// It is true if the store changed, or if a Kubernetes resource (e.g.
-	// Service, EndpointSlice) that is related to a Gateway API resource (e.g. Gateway, HTTPRoute) changed.
-	// It is reset to false after Process is called.
-	changed bool
+	cfg ChangeProcessorConfig
 
 	lock sync.Mutex
 }
 
 // NewChangeProcessorImpl creates a new ChangeProcessorImpl for the Gateway resource with the configured namespace name.
 func NewChangeProcessorImpl(cfg ChangeProcessorConfig) *ChangeProcessorImpl {
+	clusterStore := graph.ClusterState{
+		GatewayClasses: make(map[types.NamespacedName]*v1beta1.GatewayClass),
+		Gateways:       make(map[types.NamespacedName]*v1beta1.Gateway),
+		HTTPRoutes:     make(map[types.NamespacedName]*v1beta1.HTTPRoute),
+		Services:       make(map[types.NamespacedName]*apiv1.Service),
+	}
+
+	extractGVK := func(obj client.Object) schema.GroupVersionKind {
+		gvk, err := apiutil.GVKForObject(obj, cfg.Scheme)
+		if err != nil {
+			panic(fmt.Errorf("failed to get GVK for object %T: %w", obj, err))
+		}
+		return gvk
+	}
+
+	trackingUpdater := newChangeTrackingUpdater(
+		cfg.RelationshipCapturer,
+		extractGVK,
+		[]changeTrackingUpdaterObjectTypeCfg{
+			{
+				gvk:               extractGVK(&v1beta1.GatewayClass{}),
+				store:             newObjectStoreMapAdapter(clusterStore.GatewayClasses),
+				trackUpsertDelete: true,
+			},
+			{
+				gvk:               extractGVK(&v1beta1.Gateway{}),
+				store:             newObjectStoreMapAdapter(clusterStore.Gateways),
+				trackUpsertDelete: true,
+			},
+			{
+				gvk:               extractGVK(&v1beta1.HTTPRoute{}),
+				store:             newObjectStoreMapAdapter(clusterStore.HTTPRoutes),
+				trackUpsertDelete: true,
+			},
+			{
+				gvk:               extractGVK(&apiv1.Service{}),
+				store:             newObjectStoreMapAdapter(clusterStore.Services),
+				trackUpsertDelete: false,
+			},
+			{
+				gvk:               extractGVK(&discoveryV1.EndpointSlice{}),
+				store:             nil,
+				trackUpsertDelete: false,
+			},
+		},
+	)
+
+	updater := newValidatingUpsertUpdater(
+		trackingUpdater,
+		cfg.EventRecorder,
+		[]upsertValidatorFunc{
```

Comment:
> Why inject here? For unit testing?

This makes validatingUpsertUpdater extendable - if we need more validation logic, we inject an additional function, if we need to replace that function, we don't need to change validatingUpsertUpdater.

> Is this function tested somewhere?

Yes, this function is tested in ChangeProcessor tests, specifically in "Webhook validation cases".

>  Why make it anonymous?

My idea was to initialize the ChangeProcessor so that the list of all resources can be seen in one place -- here it is `ChangeProcessorImpl`. This way once when we add more resources, we need to touch less places.

Reply:
Will the anonymous function make it harder to debug in the case of an error? What does the stack trace look like if the function has an error?  Is the function fully covered by unit tests?

Reply:
Here is an example of the stacktrace if that function panics:
```
  [PANICKED] Test Panicked
  In [It] at: <REDACTED>/src/github.com/nginxinc/nginx-kubernetes-gateway/internal/state/change_processor.go:163 @ 04/04/23 13:37:57.082

  something went wrong

  Full Stack Trace
    github.com/nginxinc/nginx-kubernetes-gateway/internal/state.NewChangeProcessorImpl.func2({0x2519280?, 0xc000397cc0?})
    	<REDACTED>/src/github.com/nginxinc/nginx-kubernetes-gateway/internal/state/change_processor.go:163 +0xfc
    github.com/nginxinc/nginx-kubernetes-gateway/internal/state.(*validatingUpsertUpdater).Upsert(0xc0004191d0, {0x2519280?, 0xc000397cc0})
    	<REDACTED>/src/github.com/nginxinc/nginx-kubernetes-gateway/internal/state/store.go:261 +0x55
    github.com/nginxinc/nginx-kubernetes-gateway/internal/state.(*ChangeProcessorImpl).CaptureUpsertChange(0x0?, {0x2519280?, 0xc000397cc0?})
    	<REDACTED>/src/github.com/nginxinc/nginx-kubernetes-gateway/internal/state/change_processor.go:193 +0xb6
    github.com/nginxinc/nginx-kubernetes-gateway/internal/state_test.glob..func1.3.4()
    	<REDACTED>/src/github.com/nginxinc/nginx-kubernetes-gateway/internal/state/change_processor_test.go:2010 +0x64
```

I don't think it makes it harder to debug. 

> Is the function fully covered by unit tests?

Yes, this function is tested in ChangeProcessor tests, specifically in "Webhook validation cases".

Reply:
Yep, looks alright. Thanks for checking!